### ISSUE TITLE
Regex lookbehind failure after an (if)then|else in perl 5.10

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9555]
Migrated from [rt.perl.org#60344](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=60344) (status was 'resolved')

Searchable as RT60344$

**Comment by p5pRT** [comment]
### From @lespea

#### Created by @lespea

This is a bug report for perl from lespea@&#8203;gmail\.com\,
generated with the help of perlbug 1\.36 running under perl 5\.10\.0\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
When using any type of lookbehind after an \(if\)then|else\, the lookbehind
fails in perl 5\.10\.

I have verified this using ActiveState's perl versions 5\.8\.8 and 5\.10\.0 as
well as red hat's distributions of perl \(both 5\.8 and 5\.10\)\.

A simple test case listed below&#8203;:

perl \-e '$t = q\{sql\_processed\.csv\}; print qq\{broken\\n\}  if  $t =~
m\{\\A\(?\(?=db2\)db2|\\D\+\)\(?\<\!processed\)\\\.csv\\z\}xms;'

5\.8  OUTPUT = q\{\}
5\.10 OUTPUT = q\{broken\}

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.10.0:

Configured by SYSTEM at Tue May 13 16:52:25 2008.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=MSWin32, osvers=5.00, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -GF -W3 -MD -Zi -DNDEBUG -O1 -DWIN32
-D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT -DUSE_SITECUSTOMIZE
-DPRIVLIB_LAST_IN_INC -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS
-DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='12.00.8804', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf
-libpath:"C:\Perl\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib
uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib
msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib
uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib
msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=true, libperl=perl510.lib
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug
-opt:ref,icf  -libpath:"C:\Perl\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    33741 avoids segfaults invoking S_raise_signal() (on Linux)
    33763 Win32 process ids can have more than 16 bits
    32809 Load 'loadable object' with non-default file extension
    32728 64-bit fix for Time::Local


@INC for perl 5.10.0:
    C:/Perl/site/lib
    C:/Perl/lib
    .


Environment for perl 5.10.0:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=C:\Program Files\ActiveState Komodo Edit 5\;C:\Program
Files\Java\jre1.6.0_07\bin;C:\oracle\ora92\bin\;C:\Program
Files\Oracle\jre\1.3.1\bin\;C:\Program
Files\Oracle\jre\1.1.8\bin\;C:\Perl\site\bin;C:\Perl\bin;C:\WINNT\system32;C:\WINNT;C:\WINNT\System32\Wbem;C:\Program
Files\Diskeeper Corporation\Diskeeper\;C:\Program
Files\Hewlett-Packard\OpenView\service desk 4.5\client\bin;C:\Program
Files\IDM Computer
Solutions\UltraEdit-32\;C:\WINNT\system32\WindowsPowerShell\v1.0
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Tue\, 04 Nov 2008 12&#8203;:19&#8203;:48 \-0800\, Adam \(via RT\) \<perlbug\-followup@&#8203;perl\.org> said&#8203;:

&nbsp;  > perl \-e '$t = q\{sql\_processed\.csv\}; print qq\{broken\\n\}  if  $t =~
&nbsp;  > m\{\\A\(?\(?=db2\)db2|\\D\+\)\(?\<\!processed\)\\\.csv\\z\}xms;'

binsearch blames \.\.\.

Change 29923 by nicholas@&#8203;entropy on 2007/01/22 19&#8203;:39&#8203;:28

&nbsp;    Subject&#8203;: \[PATCH\] Disable positive lookaround optimisations
&nbsp;    From&#8203;: demerphq \<demerphq@&#8203;gmail\.com>
&nbsp;    Message\-ID&#8203;: \<9b18b3110701210953l4df6198re36a9342e6049583@&#8203;mail\.gmail\.com>
&nbsp;    Date&#8203;: Sun\, 21 Jan 2007 18&#8203;:53&#8203;:38 \+0100

Confirmed also for bleadperl

\-\- 
andreas

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

2008/11/6 Andreas J\. Koenig \<andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de>&#8203;:
>>>>>> On Tue\, 04 Nov 2008 12&#8203;:19&#8203;:48 \-0800\, Adam \(via RT\) \<perlbug\-followup@&#8203;perl\.org> said&#8203;:
>
>  > perl \-e '$t = q\{sql\_processed\.csv\}; print qq\{broken\\n\}  if  $t =~
>  > m\{\\A\(?\(?=db2\)db2|\\D\+\)\(?\<\!processed\)\\\.csv\\z\}xms;'
>
> binsearch blames \.\.\.
>
> Change 29923 by nicholas@&#8203;entropy on 2007/01/22 19&#8203;:39&#8203;:28
>
>        Subject&#8203;: \[PATCH\] Disable positive lookaround optimisations
>        From&#8203;: demerphq \<demerphq@&#8203;gmail\.com>
>        Message\-ID&#8203;: \<9b18b3110701210953l4df6198re36a9342e6049583@&#8203;mail\.gmail\.com>
>        Date&#8203;: Sun\, 21 Jan 2007 18&#8203;:53&#8203;:38 \+0100
>
> Confirmed also for bleadperl

Can you rescan please? I have a feeling this was a volatile period\,
and that in fact this patch is only incidentally related\. I think in
fact it was something to do with the de\-recursion stuff that Dave M
did\.

Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

2008/11/4 via RT Adam \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> \# New Ticket Created by  Adam
> \# Please include the string&#8203;:  \[perl \#60344\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=60344 >
>
>
> This is a bug report for perl from lespea@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.36 running under perl 5\.10\.0\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> When using any type of lookbehind after an \(if\)then|else\, the lookbehind
> fails in perl 5\.10\.
>
> I have verified this using ActiveState's perl versions 5\.8\.8 and 5\.10\.0 as
> well as red hat's distributions of perl \(both 5\.8 and 5\.10\)\.
>
> A simple test case listed below&#8203;:
>
> perl \-e '$t = q\{sql\_processed\.csv\}; print qq\{broken\\n\}  if  $t =~
> m\{\\A\(?\(?=db2\)db2|\\D\+\)\(?\<\!processed\)\\\.csv\\z\}xms;'
>
> 5\.8  OUTPUT = q\{\}
> 5\.10 OUTPUT = q\{broken\}

Hi\, Thanks for your report\.

This has been fixed with change \#34746 in bleadperl\.

http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/p/34746

Dave&#8203;: could you review this patch for sanity? It passes all tests\, and
makes sense to me\, but the de\-recursivization logic is a touch hairy
so I would appreciate a second opinion\.

Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

2008/11/6 demerphq \<demerphq@&#8203;gmail\.com>&#8203;:
> 2008/11/4 via RT Adam \<perlbug\-followup@&#8203;perl\.org>&#8203;:
>> \# New Ticket Created by  Adam
>> \# Please include the string&#8203;:  \[perl \#60344\]
>> \# in the subject line of all future correspondence about this issue\.
>> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=60344 >
>>
>>
>> This is a bug report for perl from lespea@&#8203;gmail\.com\,
>> generated with the help of perlbug 1\.36 running under perl 5\.10\.0\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> When using any type of lookbehind after an \(if\)then|else\, the lookbehind
>> fails in perl 5\.10\.
>>
>> I have verified this using ActiveState's perl versions 5\.8\.8 and 5\.10\.0 as
>> well as red hat's distributions of perl \(both 5\.8 and 5\.10\)\.
>>
>> A simple test case listed below&#8203;:
>>
>> perl \-e '$t = q\{sql\_processed\.csv\}; print qq\{broken\\n\}  if  $t =~
>> m\{\\A\(?\(?=db2\)db2|\\D\+\)\(?\<\!processed\)\\\.csv\\z\}xms;'
>>
>> 5\.8  OUTPUT = q\{\}
>> 5\.10 OUTPUT = q\{broken\}
>
> Hi\, Thanks for your report\.
>
> This has been fixed with change \#34746 in bleadperl\.
>
> http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/p/34746
>
> Dave&#8203;: could you review this patch for sanity? It passes all tests\, and
> makes sense to me\, but the de\-recursivization logic is a touch hairy
> so I would appreciate a second opinion\.

I dont seem to be able to close Perl bugs anymore\. Could someone close
this for me\.

Cheers\,
yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

This has been fixed with change \#34746 in bleadperl\.

http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/p/34746

And the ticket has been marked as resolved\.

Thanks for your report\.

yves

**Comment by p5pRT** [comment]
@demerphq - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Nov 06\, 2008 at 11&#8203;:46&#8203;:56AM \+0100\, demerphq wrote&#8203;:
> > A simple test case listed below&#8203;:
> >
> > perl \-e '$t = q\{sql\_processed\.csv\}; print qq\{broken\\n\}  if  $t =~
> > m\{\\A\(?\(?=db2\)db2|\\D\+\)\(?\<\!processed\)\\\.csv\\z\}xms;'
> >
> > 5\.8  OUTPUT = q\{\}
> > 5\.10 OUTPUT = q\{broken\}
> 
> Hi\, Thanks for your report\.
> 
> This has been fixed with change \#34746 in bleadperl\.
> 
> http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/p/34746
> 
> Dave&#8203;: could you review this patch for sanity? It passes all tests\, and
> makes sense to me\, but the de\-recursivization logic is a touch hairy
> so I would appreciate a second opinion\.

Looks good\. Thanks for fixing it\!

\-\- 
That he said that that that that is is is debatable\, is debatable\.
